src/Common/ThreadPool.cpp

```diff
@@ -230,48 +289,58 @@ ReturnType ThreadPoolImpl<Thread>::scheduleImpl(Job job, Priority priority, std:
         if (shutdown)
             return on_error("shutdown");
 
+        bool adding_new_thread = new_thread && threads.size() < std::min(max_threads, 1 + scheduled_jobs + max_free_threads);
+        typename std::list<std::unique_ptr<ThreadFromThreadPool>>::iterator thread_slot;
+
         /// We must not to allocate any memory after we emplaced a job in a queue.
         /// Because if an exception would be thrown, we won't notify a thread about job occurrence.
-
-        /// Check if there are enough threads to process job.
-        if (threads.size() < std::min(max_threads, scheduled_jobs + 1))
+        if (adding_new_thread)
         {
+            // we've got the new thread, it was started out of the locked section,
+            // so in between the demand on thread may have been changed.
+            //
+            // Even if the demand can be lower already but if the pool can still take
+            // new thread - let's add it since we already have it.
```

Comment:
TODO: need to check if the opposite is possible (when outside of the critical section we decided to NOT create the thread, but now we see it's needed). Check the order of increments / decrements and may be add some fallback.

Reply:
Fixed.